internal/controller/managedcluster_controller.go

```diff
@@ -404,6 +440,90 @@ func (r *ManagedClusterReconciler) Delete(ctx context.Context, l logr.Logger, ma
 	return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 }
 
+func (r *ManagedClusterReconciler) releaseCluster(ctx context.Context, namespace, clusterName, templateName string) error {
+	providers, err := r.getProviders(ctx, templateName)
+	if err != nil {
+		return err
+	}
+
+	providerGVKs := map[string]providerSchema{
+		"aws":   {machine: gvkAWSMachine, cluster: gvkAWSCluster},
+		"azure": {machine: gvkAzureMachine, cluster: gvkAzureCluster},
+	}
+
+	// Associate the provider with it's GVK
+	for _, provider := range providers {
+		gvk, ok := providerGVKs[provider]
+		if !ok {
+			continue
+		}
+
+		found, err := r.machinesAvailable(ctx, namespace, clusterName, gvk.machine)
```

Comment:
Same situation here, managedCluster.Name != cluster.Name

Reply:
I did some experiments and found the following:
Machine, AWSMachine or AzureMachine objects don't have labels like "helm.toolkit.fluxcd.io/name" (which can be used to filter the objects by managed cluster name), e.g. they have these labels only:
```labels:
    cluster.x-k8s.io/cluster-name: aws-dev-XXXXX
    cluster.x-k8s.io/control-plane: ""
    cluster.x-k8s.io/control-plane-name: aws-dev-XXXXX-cp
```
MachineDeployment object has "helm.toolkit.fluxcd.io/name", but MachineDeployment is removed immediately when Helm Release started deletion process (but Machines are still in the process of removal).
Should we apply labels on Machines during reconciliation and then use such labels watch them?

Reply:
You can get the name of the cluster using the flux label. Then use this name to query all associated machines.